### PR TITLE
Squeeze down the size of the skeleton reference cache.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.SkeletonReferenceCache.cs [hunk 1]
```diff
@@ -213,23 +215,6 @@ public SkeletonReferenceCache Clone()
...
-                if (version == _version)
-                {
-                    result = _skeletonReferenceSet;
-                    return true;
-                }
-            }
-
-            result = null;
-            return false;
-        }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.SkeletonReferenceCache.cs:231]: inlined this method.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.SkeletonReferenceCache.cs [hunk 2]
```diff
@@ -78,54 +85,40 @@ private partial class SkeletonReferenceCache
...
-        {
-        }
-
-        private SkeletonReferenceCache(
-            VersionStamp? version,
-            SkeletonReferenceSet? skeletonReferenceSet)
-        {
-            _version = version;
-            _skeletonReferenceSet = skeletonReferenceSet;
-        }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.SkeletonReferenceCache.cs:108]: don't need/want the constructors.  The default state of this struct is correct and what we want.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.SkeletonReferenceCache.cs [hunk 3]
```diff
@@ -57,10 +56,18 @@ internal partial class SolutionCompilationState
     /// memory.
     /// </para>
     /// </summary>
-    private partial class SkeletonReferenceCache
+    [NonCopyable]
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.SkeletonReferenceCache.cs:59]: prevents accidental copies, which could produce torn views of the struct.  Clone is the only supported mechanism to get another copy.  will doc.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 4]
```diff
@@ -682,14 +685,14 @@ public ICompilationTracker FreezePartialState(CancellationToken cancellationToke
             {
                 var state = this.ReadState();
 
-                var clonedCache = this.SkeletonReferenceCache.Clone();
+                var clonedSkeletonReferenceCache = _skeletonReferenceCache.Clone();
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:688]: 💡 non-copyable issue should be resolved by inlining this call to clone at the use sites of this local variable

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 5]
```diff
@@ -131,7 +134,7 @@
                 return new CompilationTracker(
                     newProjectState,
                     forkedTrackerState,
-                    this.SkeletonReferenceCache.Clone());
+                    _skeletonReferenceCache.Clone());
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:140]: >_skeletonReferenceCache.Clone [](http://example.com/codeflow?start=20&length=29)

nit: maybe use GetClonedSkeletonReferenceCache here and in FreezePartialState

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.GeneratedFileReplacingCompilationTracker.cs [hunk 6]
```diff
@@ -19,8 +18,10 @@ internal partial class SolutionCompilationState
     /// to return a generated document with a specific content, regardless of what the generator actually produces. In other words, it says
     /// "take the compilation this other thing produced, and pretend the generator gave this content, even if it wouldn't."
     /// </summary>
-    private class GeneratedFileReplacingCompilationTracker(ICompilationTracker underlyingTracker, TextDocumentStates<SourceGeneratedDocumentState> replacementDocumentStates) : ICompilationTracker
+    private class GeneratedFileReplacingCompilationTracker : ICompilationTracker
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.GeneratedFileReplacingCompilationTracker.cs:21]: had to move off of primary constructor as the non-copyable analyzer was throwing up false positives with the initialization of the SkeletonReferenceCache  field.

**Comment by CyrusNajmabadi** [comment]
Test insertion: https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/534024

**Comment by ToddGrun** [comment]
Verified in the test insertion that allocations under CompilationTracker.Fork related to SkeletonReferenceCache are reduced

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski 	For review when you get back.

**Pull request merged**